src/groupby.jl

```diff
@@ -0,0 +1,341 @@
+"""
+    DimGroupByArray <: AbstractDimArray
+
+`DimGroupByArray` is essentially a `DimArray` but holding
+the results of a `groupby` operation.
+
+Its dimensions are the sorted results of the grouping functions
+used in `groupby`.
+
+This wrapper allows for specialisations on later broadcast or
+reducing operations, e.g. for chunk reading with DiskArrays.jl,
+because we know the data originates from a single array.
+"""
+struct DimGroupByArray{T,N,D<:Tuple,R<:Tuple,A<:AbstractArray{T,N},Na,Me} <: AbstractDimArray{T,N,D,A}
+    data::A
+    dims::D
+    refdims::R
+    name::Na
+    metadata::Me
+    function DimGroupByArray(
+        data::A, dims::D, refdims::R, name::Na, metadata::Me
+    ) where {D<:Tuple,R<:Tuple,A<:AbstractArray{T,N},Na,Me} where {T,N}
+        checkdims(data, dims)
+        new{T,N,D,R,A,Na,Me}(data, dims, refdims, name, metadata)
+    end
+end
+function DimGroupByArray(data::AbstractArray, dims::Union{Tuple,NamedTuple};
```

Comment:
This is great. I think with this DimGroupByArray we should have everything we need to make the DiskArrayEngine Wrapper